src/store/matchedAudiences.ts

```diff
@@ -30,23 +39,74 @@ class MatchedAudienceStore {
     this._save();
   }
 
-  _save() {
+  _save(): void {
     storage.set(StorageKeys.MATCHED_AUDIENCES, this.matchedAudiences);
     storage.set(StorageKeys.MATCHED_AUDIENCE_IDS, this.matchedAudienceIds);
   }
 
-  setMatchedAudiences(newlyMatchedAudiences: MatchedAudience[]) {
-    // TODO: decide if we need to check duplicate audiences here...
+  setMatchedAudiences(matchedAudiences: MatchedAudience[]): void {
+    this.matchedAudiences = matchedAudiences;
+    this.matchedAudienceIds = matchedAudiences.map((audience) => audience.id);
+  }
 
-    if (newlyMatchedAudiences.length === 0) return;
-    const allMatchedAudiences = [
-      ...this.matchedAudiences,
-      ...newlyMatchedAudiences,
-    ];
-    this.matchedAudiences = allMatchedAudiences;
-    this.matchedAudienceIds = allMatchedAudiences.map(
-      (audience) => audience.id
+  getMatchedAudiences(): MatchedAudience[] {
+    return [...this.matchedAudiences];
+  }
+
+  saveMatchingAudiences(
+    audienceDefinitions: AudienceDefinition[],
+    viewStore: ViewStore
+  ): void {
+    const currentTS = timeStampInSecs();
+
+    const newlyMatchedAudiences = audienceDefinitions
+      // drop prev matched audiences with same version
+      .filter((audience) => {
+        return !this.matchedAudiences.some((matchedAudience) =>
+          doesAudienceVersionMatch(audience, matchedAudience)
+        );
+      })
+      // translate audience definitions into engine queries
+      .map((audience) => {
+        return {
+          ...audience,
+          conditions: engine.translate(audience),
+        };
+      })
+      // check if query matches any pageViews
+      .map((query) => {
+        // TODO Ideally pageViewsWithinLookBack should be passed as
+        // argument to saveMatchingAudiences methods, but as it is now
+        // we got to take the entire viewStore instance as arg,
+        // or else there will be repetition.
+        const pageViewsWithinLookBack = viewStore.getPageViewsWithinLookBack(
+          query.lookBack
+        );
```

Comment:
The coupling is caused by this.
As it is, `viewStore.getPageViewsWithinLookBack` needs que `lookBack` param for each `audienceDefinition`.